src/libraries/Common/src/System/TimeProvider.cs

```diff
@@ -0,0 +1,209 @@
+        /// but can be re-enabled using the <see cref="ITimer.Change"/> method.
+        /// </para>
+        /// <para>
+        /// The return <see cref="ITimer"/> instance will be implicitly rooted while the timer is still scheduled.
+        /// </para>
+        /// <para>
+        /// <see cref="CreateTimer"/> captures the <see cref="ExecutionContext"/> and stores that with the <see cref="ITimer"/> for use in invoking <paramref name="callback"/>
+        /// each time it's called. That capture can be suppressed with <see cref="ExecutionContext.SuppressFlow"/>.
+        /// </para>
+        /// </remarks>
+        public abstract ITimer CreateTimer(TimerCallback callback, object? state, TimeSpan dueTime, TimeSpan period);
+
+        /// <summary>
+        /// Provides a default implementation of <see cref="TimeProvider"/> based on <see cref="DateTimeOffset.UtcNow"/>,
+        /// <see cref="TimeZoneInfo.Local"/>, <see cref="Stopwatch"/>, and <see cref="Timer"/>.
+        /// </summary>
+        private sealed class SystemTimeProvider : TimeProvider
+        {
+            /// <summary>The time zone to treat as local.  If null, <see cref="TimeZoneInfo.Local"/> is used.</summary>
+            private readonly TimeZoneInfo? _localTimeZone;
+
+            /// <summary>Initializes the instance.</summary>
+            /// <param name="localTimeZone">The time zone to treat as local.  If null, <see cref="TimeZoneInfo.Local"/> is used.</param>
+            internal SystemTimeProvider(TimeZoneInfo? localTimeZone) : base(Stopwatch.Frequency) => _localTimeZone = localTimeZone;
+
+            /// <inheritdoc/>
+            public override TimeZoneInfo LocalTimeZone => _localTimeZone ?? TimeZoneInfo.Local;
+
+            /// <inheritdoc/>
+            public override ITimer CreateTimer(TimerCallback callback, object? state, TimeSpan dueTime, TimeSpan period)
+            {
+                ArgumentNullException.ThrowIfNull(callback);
+                return new SystemTimeProviderTimer(dueTime, period, callback, state);
+            }
+
+            /// <inheritdoc/>
+            public override long GetTimestamp() => Stopwatch.GetTimestamp();
+
+            /// <inheritdoc/>
+            public override DateTimeOffset UtcNow => DateTimeOffset.UtcNow;
+
+            /// <summary>Thin wrapper for a <see cref="TimerQueueTimer"/>.</summary>
+            /// <remarks>
+            /// We don't return a TimerQueueTimer directly as it implements IThreadPoolWorkItem and we don't
+            /// want it exposed in a way that user code could directly queue the timer to the thread pool.
+            /// We also use this instead of Timer because CreateTimer needs to return a timer that's implicitly
+            /// rooted while scheduled.
+            /// </remarks>
+            private sealed class SystemTimeProviderTimer : ITimer
+            {
+                private const uint MaxSupportedTimeout = 0xfffffffe;
```

Comment:
Can the below just use Timer.MaxSupportedTimeout instead of duplicating the constant value here?  Any limitation in this class stems from the maximum supported by `Timer`, so it makes sense that we'd be referring to its limit.

Reply:
We can use the `Timer.MaxSupportedTimeout` but I am planning to use the same code when implementing the down-level version which will not have access to it. I'll use `Timer.MaxSupportedTimeout` for now and then I may consider ifdef or rename it later.